storage/src/sparse_merkle/tree.cpp

```diff
@@ -0,0 +1,198 @@
+// Concord
+//
+// Copyright (c) 2019 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the
+// LICENSE file.
+
+#include "sparse_merkle/tree.h"
+#include "hash_defs.h"
+
+using namespace concordUtils;
+
+namespace concord::storage::sparse_merkle {
+
+// All data that is loaded from the DB, manipulated, and/or returned from a
+// call to `update`.
+//
+// A new instance of this structure is created during every update call.
+struct UpdateCache {
```

Comment:
I think it's better to have those structures to be defined in the include file.

Reply:
I originally had them in there, but it presented a few problems. For example, since they were private, internally linked (plain functions) in the CPP file could not use them. Making them public or putting them in the namespace, also effectively makes them public and part of the API. They are both purely implementation details so I put them in the cpp file. They are both subject to change with future optimzations.

Reply:
I'm also open to adding an "implementation class" in a future PR if necessary. That would allow direct testing of these helper functions.